head_first_design_patterns/tbruinem/json/src/include/parser.hpp

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <cstdint>
+#include <cstddef>
+
+#include "common/string.hpp"
+#include "element_type.hpp"
+
+namespace json {
+
+class ParsedJSON;
+
+enum class ParseState {
+	OUT,	//! Initial state
+	WORD,	//! Name of a JSON item
+	VALUE,	//! Value of a JSON item
+	IN,		//! Between Name and Value
+};
+
+enum class StateResult {
+	CONTINUE,
+	DONE,
+	ERROR
+};
+
+class Parser {
+public:
+	Parser(string::iterator& it, const string::iterator& end, ElementType parent = ElementType::JSON_ARRAY) :
```

Comment:
I feel like it would make more sense for a json parser to be a function as opposed to being a class